### PR TITLE
pythonPackages: new and updates

### PR DESCRIPTION
**Pull request by oxij**
###### Motivation for this change
"Because YouTube gone complete JavaScript"

lz4: unrelated update, but I'm too lazy to make a separate PR for that.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by FRidh — CHANGES_REQUESTED** [review]
Can you move the expressions into their own files (see the header of `python-packages.nix`) and use `fetchPypi` where possible.

**Comment by FRidh** [comment]
I will eventually add your PR to #29009.

**Comment by oxij** [comment]
Frederik Rietdijk <notifications@github.com> writes:

> Can you move the expressions into their own files (see the header of `python-packages.nix`) and use `fetchPypi` where possible.

Done.

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -1730,6 +1730,8 @@ in {
     };
   };
 
+  comp = callPackage ../applications/video/comp { };
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: This is an application and should therefore be called directly from `all-packages.nix`, and not via `python-packages.nix`.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -23813,6 +23782,8 @@ EOF
 
   wheel = callPackage ../development/python-modules/wheel { };
 
+  whitey = callPackage ../applications/video/whitey { };
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: This is an application and should therefore be called directly from `all-packages.nix`, and not via `python-packages.nix`.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -26812,39 +26763,7 @@ EOF
...
-    propagatedBuildInputs = with self; [ pafy ];
-
-    meta = with stdenv.lib; {
-      description = "Terminal based YouTube player and downloader";
-      homepage = http://github.com/np1/mps-youtube;
-      license = licenses.gpl3;
-      maintainers = with maintainers; [ odi ];
-    };
-  };
+  mps-youtube = callPackage ../applications/video/mps-youtube { };
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: This is an application and should therefore be called directly from `all-packages.nix`, and not via `python-packages.nix`.

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by oxij** [comment]
> This is an application and should therefore be called directly from `all-packages.nix`, and not via `python-packages.nix`.

I respectfully disagree with that policy. I think things that use
language-specific build system should live near said language-specific
build system. Especially when autogeneration of those package
expressions is possible (and it clearly is possible for most Python
packages present in PyPi, it's just nobody got to implementing it yet).

Haskell, Emacs, Perl, etc use the opposite policy and refer to
applications from `all-packages.nix` with `*Packages.?*.package` when
needed (see, for instance, `git-annex`).

* Why does this policy for Python differs from policies used for
  Haskell, Emacs, Perl, etc?

It also makes little sense to move Python applications to
`all-packages.nix` for technical reasons:

* In `python-packages.nix` `callPackage` substitutes python deps, python
  programs use a lot of python deps. So it only makes sense to leave
  related packages there.

* For Python application that can work under different interpreter
  versions you have to select a single interpreter version in
  `all-packages.nix`. No such problem in `python-packages.nix`.

* You can always refer to any package from `all-packages.nix` with e.g.
  `mps-youtube = python3Packages.mps-youtube` if needed.

To be perfectly honest, I would prefer Python packages to live in a
single file sorted by attribute name, but technical arguments for that
get less and less valid with better SSDs and bigger RAM (unless the
included files are so tiny that `open(2)` costs much more than `read(2)`
followed by parsing).

**Comment by FRidh** [comment]
> I respectfully disagree with that policy. I think things that use
language-specific build system should live near said language-specific
build system.

Where do packages live that are written in multiple languages?

> Especially when autogeneration of those package
expressions is possible (and it clearly is possible for most Python
packages present in PyPi, it's just nobody got to implementing it yet).

Unfortunately it is *not* possible with Python because:

1. dependencies are not declarative;
2. there is no curated list of packages that are known/solved to work together.

`pypi2nix` is a nice attempt, circumventing 1) by simply building everything. Distutils discussed creating a solver for PyPI, but that won't happen anytime soon. Furthermore, as long as its not possible to query PyPI at a certain state there is no way to create package sets that "match" for the different interpreter version/platform combinations that we have.

> To be perfectly honest, I would prefer Python packages to live in a
single file sorted by attribute name, but technical arguments for that
get less and less valid with better SSDs and bigger RAM (unless the
included files are so tiny that `open(2)` costs much more than `read(2)`
followed by parsing).

That's how it was, and how I liked it as well, but we're moving away from that because its problematic, e.g. due to merge conflicts. If we could have autogenerated data, then having a single file with overrides would be nice. But that's not going to happen anytime soon.

**Comment by oxij** [comment]
Note that while this PR bitrotted much of _the exactly the original changes from this PR_ were merged into master without any discussion on "policy". Doesn't that look hypocritical to you?

**Comment by 7c6f434c** [comment]
>Where do packages live that are written in multiple languages?

They don't use the stereotypical package-constructing function of any original languagePackages anyway.

**Pull request closed without merging**